src/relay/ir/text_printer.cc

```diff
@@ -0,0 +1,752 @@
+    this->PrintEndInst("\n");
+    return val;
+  }
+
+  TextValue VisitExpr_(const GlobalVarNode* op) final {
+    return TextValue('@' + op->name_hint);
+  }
+
+  TextValue VisitExpr_(const FunctionNode* op) final {
+    TextValue id = AllocTempVar();
+    std::ostringstream os;
+    os << id << " = function";
+    this->PrintFuncInternal(os.str(), GetRef<Function>(op));
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const CallNode* op) final {
+    // TODO(tqchen, marisha): support generic call
+    // possibly through meta-data
+    CHECK_EQ(op->type_args.size(), 0U)
+        << "generic call not yet supported";
+    TextValue call_op = GetValue(op->op);
+    std::vector<TextValue> args;
+    for (Expr arg : op->args) {
+      args.emplace_back(GetValue(arg));
+    }
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = " << call_op << "(";
+    for (size_t i = 0; i < args.size(); ++i) {
+      stream_ << args[i];
+      if (i + 1 != args.size()) {
+        stream_ << ", ";
+      }
+    }
+    this->PrintCallAttrs(op->op, op->attrs, stream_);
+    stream_ << ")";
+    this->PrintEndInst("");
+    this->PrintOptionalInfo(GetRef<Expr>(op));
+    stream_ << '\n';
+    return id;
+  }
+
+  TextValue VisitExpr_(const LetNode* op) final {
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = ";
+    this->PrintScope(GetRef<Expr>(op));
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const IfNode* op) final {
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = ";
+    this->PrintScope(GetRef<Expr>(op));
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const OpNode* op) final {
+    return TextValue(op->name);
+  }
+
+  TextValue VisitExpr_(const TupleGetItemNode* op) final {
+    TextValue tuple = GetValue(op->tuple);
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = " << tuple << "[" << op->index << "]";
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  /*!
+   * \brief Print the type to os
+   * \param type The type to be printed.
+   * \param os The output type.
+   */
+  void PrintType(const Type& type, std::ostream& os) {  // NOLINT(*)
+    this->VisitType(type, os);
+  }
+  //------------------------------------
+  // Overload of Expr printing functions
+  //------------------------------------
+  void VisitType_(const TensorTypeNode* node, std::ostream& os) final {  // NOLINT(*)
+    // scalar type
+    if (node->shape.size() == 0) {
+      os << runtime::TVMType2String(Type2TVMType(node->dtype));
+      return;
+    }
+    os << "Tensor[(";
+    for (size_t i = 0; i < node->shape.size(); ++i) {
+      this->PrintAttr(node->shape[i], os);
+      if (i + 1 != node->shape.size()) {
+        os << ", ";
+      }
+    }
+    if (node->shape.size() == 1) {
+      os << ",";
```

Comment:
Seems no need to print `,` for size of 1, right?

Reply:
It conforms to what people are used to from Python, I imagine

Reply:
It conforms to the python tuple standard.